base/pmap.jl

```
@@ -15,38 +19,170 @@ Note that `f` must be made available to all worker processes; see
 for details.
+
+If a worker pool is not specified, all available workers, i.e., the default worker pool
+is used.
+
+By default, `pmap` distributes the computation over all specified workers. To use only the
+local process and distribute over tasks, specify `distributed=false`. This is equivalent to `asyncmap`.
+
+`pmap` can also use a mix of processes and tasks via the `batch_size` argument. For batch sizes
+greater than 1, the collection is split into multiple batches, which are distributed across
+workers. Each such batch is processed in parallel via tasks in each worker. The specified
+`batch_size` is an upper limit, the actual size of batches may be smaller and is calculated
+depending on the number of workers available and length of the collection.
+
+Any error stops pmap from processing the remainder of the collection. To override this behavior
+you can specify an error handling function via argument `on_error` which takes in a single argument, i.e.,
+the exception. The function can stop the processing by rethrowing the error, or, to continue, return any value
+which is then returned inline with the results to the caller.
+
+Failed computation can also be retried via `retry_on`, `retry_n`, `retry_max_delay`, which are passed through
+to `retry` as arguments `retry_on`, `n` and `max_delay` respectively. If batching is specified, and an entire batch fails,
+all items in the batch are retried.
+
+The following are equivalent:
+
+* `pmap(f, c; distributed=false)` and `asyncmap(f,c)`
+* `pmap(f, c; retry_n=1)` and `asyncmap(retry(remote(f)),c)`
+* `pmap(f, c; retry_n=1, on_error=e->e)` and `asyncmap(x->try retry(remote(f))(x) catch e; e end, c)`
 """
-pmap(p::WorkerPool, f, c...) = collect(pgenerate(p, f, c...))
+function pmap(p::WorkerPool, f, c...; kwargs...)
+    results_iter, process_errors! = pgenerate(p, f, c...; kwargs...)
+    results = collect(results_iter)
+    if isa(process_errors!, Function)
+        process_errors!(p, f, results)
+    end
```

Should there be an `else` here? or should it be an assertion? If `process_errors` is not a Function, it is silently ignored.